src/server.c

```diff
@@ -4085,25 +4085,50 @@ uint64_t getCommandFlags(client *c) {
     return cmd_flags;
 }
 
-/* Prepare a parsed command for processing, including looking up the command,
- * checking arity and calculating cluster slot. This should be done before
- * calling processCommand() and can be done by I/O threads to offload the
+/* Prepare the client's current command for processing, including looking up the
+ * command, checking arity and calculating cluster slot. This should be done
+ * before calling processCommand() and can be done by I/O threads to offload the
  * main-thread. */
 void prepareCommand(client *c) {
+    if (!(c->read_flags & READ_FLAGS_PARSING_COMPLETED) || c->argc == 0) return;
     /* Make sure we don't do this twice. */
     debugServerAssert(c->parsed_cmd == NULL && !(c->read_flags & READ_FLAGS_COMMAND_NOT_FOUND));
     c->parsed_cmd = lookupCommand(c->argv, c->argc);
-    if (c->parsed_cmd && !commandCheckArity(c->parsed_cmd, c->argc, NULL)) {
-        /* The command was found, but the arity is invalid. */
+    if (!c->parsed_cmd) {
+        c->read_flags |= READ_FLAGS_COMMAND_NOT_FOUND;
+    } else if (!commandCheckArity(c->parsed_cmd, c->argc, NULL)) {
         c->read_flags |= READ_FLAGS_BAD_ARITY;
-    } else if (c->parsed_cmd && server.cluster_enabled) {
+    } else if (server.cluster_enabled) {
         debugServerAssert(c->slot == -1 &&
                           !(c->read_flags & READ_FLAGS_CROSSSLOT) &&
                           !(c->read_flags & READ_FLAGS_NO_KEYS));
         c->slot = clusterSlotByCommand(c->parsed_cmd, c->argv, c->argc, &c->read_flags);
     }
 }
 
+/* Prepare all parsed commands in the client's queue. See prepareCommand(). */
+void prepareCommandQueue(client *c) {
```

Comment:
It would be nice if this function called prepareCommand for both the main command as well as the ones in the command queue, instead of having all the code duplicated.

Reply:
Added a prepareCommandGeneric that works on pointers to argv, argc, read_flags, cmd and slot.